
<a href="https://colab.research.google.com/github/aleylani/Databehandling-AI24/blob/main/exercises/E3_merging.ipynb" target="_parent"><img align="left" src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a> &nbsp; to see hints and answers.

---
# Merging exercises

---
These are introductory exercises in Pandas with focus in **syntax, indexing, data selection, missing data, aggregations, visualizations**, **data cleaning**, **merging**, **concatenation**, **joining**, **parse html tables** .

<p class = "alert alert-info" role="alert"><b>Note</b> all datasets used in this exercise can be found under Data folder of the course Github repo</p>

<p class = "alert alert-info" role="alert"><b>Note</b> that there are usually many ways to achieve same results using operations in pandas. Feel free to investigate several ways. </p>

<p class = "alert alert-info" role="alert"><b>Note</b> that in cases when you start to repeat code, try not to. Create functions to reuse code instead. </p>

<p class = "alert alert-info" role="alert"><b>Remember</b> to use <b>descriptive variable, function, index </b> and <b> column names</b> in order to get readable code </p>

<p class = "alert alert-info" role="alert"><b>Remember</b> to format your input questions in a pedagogical way to guide the user

The number of stars (\*), (\*\*), (\*\*\*) denotes the difficulty level of the task

---

**Note that you might need to install lxml in your environment**

In [57]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

----
**A)**

In [58]:
se_demo_df = pd.read_html('https://sv.wikipedia.org/wiki/Sveriges_demografi')



---
**B)**

In [137]:
df_1b = se_demo_df[6]
df_1b
#sns.histplot(data=df_1b, x="Folkmängd", kde=True, bins=30)

#plt.show()



Unnamed: 0 Medel- folkmängd    Födda    Döda  \
0          1900        5 117 000  138 139  86 146   
1          1901        5 156 000  139 370  82 772   
2          1902        5 187 000  137 364  79 722   
3          1903        5 210 000  133 896  78 610   
4          1904        5 241 000  134 952  80 152   
..          ...              ...      ...     ...   
118        2018     10 175 213,5  115 832  92 185   
119        2019     10 278 887,0  114 523  88 766   
120        2020     10 353 442,0  113 077  98 124   
121        2021     10 415 810,5  114 263  91 958   
122        2022     10 521 556,0  104 734  94 737   

    Naturlig befolknings- förändring  Nativitet (per 1000)  \
0                             51 993                 270.0   
1                             56 598                 270.0   
2                             57 642                 265.0   
3                             55 286                 257.0   
4                             54 800                 257.0   
..                               ...                   ...   
118                           23 647                 113.0   
119                           28 727                 111.0   
120                           14 953                 109.0   
121                           22 305                 110.0   
122                            9 997                   NaN   

     Dödstal (per 1000)  Naturlig befolknings- förändring (per 1000)  \
0                 168.0                                        102.0   
1                 161.0                                        110.0   
2                 154.0                                        111.0   
3                 151.0                                        106.0   
4                 153.0                                        105.0   
..                  ...                                          ...   
118                90.0                                         23.0   
119                87.0                                         24.0   
120                95.0                                         14.0   
121                88.0                                         22.0   
122                 NaN                                          NaN   

     Total fertilitet  
0               402.0  
1               404.0  
2               395.0  
3               382.0  
4               383.0  
..                ...  
118             175.0  
119             170.0  
120             166.0  
121               NaN  
122               NaN  

[123 rows x 9 columns]

----
**C)**

In [114]:
df_1c = se_demo_df[1]
df_1c2 = pd.DataFrame()
df_1c2["År"] = df_1c["Vid utgången av år"][0:7]
df_1c2["Folkmängd"] = df_1c["Folkmängd"][0:7]

df_1c2 = df_1c2.reset_index(drop=True)
df_1c2

År  Folkmängd
0  1570    900 000
1  1650  1 225 000
2  1700  1 485 000
3  1720  1 350 000
4  1755  1 878 000
5  1815  2 465 000
6  1865  4 099 000

---
**D)**

In [160]:
#df_1c2_clean.info()
df_1b_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   År         123 non-null    int64 
 1   Folkmängd  123 non-null    object
dtypes: int64(1), object(1)
memory usage: 2.1+ KB


In [ ]:

df_1b
#print(df_1c2.head())

df_1b_clean = df_1b.drop(["Födda", "Döda", "Naturlig befolknings- förändring", "Nativitet (per 1000)", "Dödstal (per 1000)", "Naturlig befolknings- förändring (per 1000)", "Total fertilitet"], axis=1)
df_1b_clean = df_1b_clean.rename(columns={"Unnamed: 0": "År", "Medel- folkmängd": "Folkmängd"})
print(df_1b_clean)


df_1c2_clean = df_1c2
df_1c2_clean["Folkmängd"] = df_1c2["Folkmängd"].astype(float)
print(df_1c2)


df_1d = df_1b_clean.merge(df_1c2,on="År")
df_1d

       År     Folkmängd
0    1900     5 117 000
1    1901     5 156 000
2    1902     5 187 000
3    1903     5 210 000
4    1904     5 241 000
..    ...           ...
118  2018  10 175 213,5
119  2019  10 278 887,0
120  2020  10 353 442,0
121  2021  10 415 810,5
122  2022  10 521 556,0

[123 rows x 2 columns]


ValueError: could not convert string to float: '900\xa0000'

## 1. Swedish demographic data (*)

Go to Swedish-language wikipedia page [Sveriges demografi](https://sv.wikipedia.org/wiki/Sveriges_demografi). 

&nbsp; a) Read in the table under "Befolkningsstatistik sedan 1900" into a DataFrame

          Note that you might need to install lxml in your environment

          conda install lxml

&nbsp; b) Choose to do some EDA (exploratory data analysis) on this dataset. And draw some relevant graphs.

&nbsp; c) Now we want to go backwards in time (before 1900) to see how population has changed in Sweden. Read in the table under history and keep the data of "Folkmängd" from 1570-1865. 

| År   | Folkmängd |
| ---- | --------- | 
| 1570 | 900000    |     
| 1650 | 1225000   |
| 1700 | 1485000   |
| 1720 | 1350000   |
| 1755 | 1878000   |
| 1815 | 2465000   |
| 1865 | 4099000   |


&nbsp; d) Now concatenate this with the table from 1900 so that you have population data from 1570 to 2020. Note that you may need to clean the data in order for it to fit properly. Also you may be able to do this in several ways. 

&nbsp; e) Draw a graph of population data from 1570-2020.

&nbsp; f) Try to see if you can plot something from this data that might show the effects of the Covid-19 pandemic. It's good to zoom in the data, i.e. plot a shorter time frame instead of all time frames. 

&nbsp; g) Try to see if you can plot something from this data that might capture the effects imigration and emigration waves.

&nbsp; h) Feel free to find other interesting facts about the Swedish demographics based on this data. 

<details>
<summary>Hint</summary>

Useful methods:
- append()
- join()
- concat()
- merge()

</details>

<br/>

<details>

<summary>Answer</summary>

![Sweden population data 1952-2020](../assets/sverige_befolkning_tid.png)

</details>

## 1. Swedish demographic data (*)

Go to Swedish-language wikipedia page [Sveriges demografi](https://sv.wikipedia.org/wiki/Sveriges_demografi). 

&nbsp; a) Read in the table under "Befolkningsstatistik sedan 1900" into a DataFrame

          Note that you might need to install lxml in your environment

          conda install lxml

&nbsp; b) Choose to do some EDA (exploratory data analysis) on this dataset. And draw some relevant graphs.

&nbsp; c) Now we want to go backwards in time (before 1900) to see how population has changed in Sweden. Read in the table under history and keep the data of "Folkmängd" from 1570-1865. 

| År   | Folkmängd |
| ---- | --------- | 
| 1570 | 900000    |     
| 1650 | 1225000   |
| 1700 | 1485000   |
| 1720 | 1350000   |
| 1755 | 1878000   |
| 1815 | 2465000   |
| 1865 | 4099000   |


&nbsp; d) Now concatenate this with the table from 1900 so that you have population data from 1570 to 2020. Note that you may need to clean the data in order for it to fit properly. Also you may be able to do this in several ways. 

&nbsp; e) Draw a graph of population data from 1570-2020.

&nbsp; f) Try to see if you can plot something from this data that might show the effects of the Covid-19 pandemic. It's good to zoom in the data, i.e. plot a shorter time frame instead of all time frames. 

&nbsp; g) Try to see if you can plot something from this data that might capture the effects imigration and emigration waves.

&nbsp; h) Feel free to find other interesting facts about the Swedish demographics based on this data. 

<details>
<summary>Hint</summary>

Useful methods:
- append()
- join()
- concat()
- merge()

</details>

<br/>

<details>

<summary>Answer</summary>

![Sweden population data 1952-2020](../assets/sverige_befolkning_tid.png)

</details>

---
## 2. Denmark demographic data (*)

Go to the Danish-language wikipedia page [Danmarks demografi](https://da.wikipedia.org/wiki/Danmarks_demografi). 

&nbsp; a) Read in the table under "Demografiske data" into a DataFrame (*)

&nbsp; b) Clean the data and draw a graph of population against year from 1769-2020. (**)


---
## 3. Norwegian demographic data (*)

Go to Swedish-language wikipedia page [Norges demografi](https://sv.wikipedia.org/wiki/Norges_demografi). 

&nbsp; a) Read in the table under "Befolkningsstatistik sedan 1900" into a DataFrame

&nbsp; b) You see some missing data in column "Total fertilitet". Go to the [English page](https://en.wikipedia.org/wiki/Demographics_of_Norway) and read in the data from "Vital statistics since 1900".  

&nbsp; c) Pick out the fertility column from b) dataset, merge it into a) dataset and clean the data so that you only have columns "År", "Folkmängd", "Fertilitet". 


---
## 4. Merge Sweden-Norway (*)

Create a population graph and a fertility graph showing Sweden and Norway.

<details>

<summary>Answer</summary>

![Fertilitet Norge och Sverige](../assets/fertilitet_sv_no.png)

![Folkmängd Norge och Sverige](../assets/folkmangd_sverige_norge.png)

</details>